Implement DeepSeek V2

EricLBuehler opened the pull request:
This PR implements the DeepSeek V2 architecture.

LaurentMazare commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1114 @@
+        };
+
+        if self.top_k > 1 && self.cfg.norm_topk_prob {
+            let denmoninator = (topk_weight.sum_keepdim(D::Minus1)? + 1e-20)?;
```
typo: denominator

EricLBuehler replied:
Thanks!

EricLBuehler commented on candle-examples/Cargo.toml:
```diff
@@ -122,3 +122,7 @@ required-features = ["onnx"]
 [[example]]
 name = "colpali"
 required-features = ["pdf2image"]
+
```
Fixed.

EricLBuehler commented on candle-examples/examples/deepseekv2/main.rs:
```diff
@@ -0,0 +1,282 @@
+    };
+    let device = candle_examples::device(args.cpu)?;
+    let (model, device) = {
+        let dtype = if device.is_cuda() {
```
Fixed, I flipped the logic so we only use `f16` on the cpu.

EricLBuehler commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1114 @@
+        // Reorder the indices ascending
+        let reorder_indices = topk_indices_sorted.arg_sort_last_dim(true)?;
+        let topk_indices_unsorted = topk_indices_sorted
+            .to_dtype(DType::F32)?
```
Fixed, this was unnecessary.

EricLBuehler commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1114 @@
+    Linear,
+}
+
+impl FromStr for ScaledRopeType {
```
Absolutely, removed.

EricLBuehler commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1114 @@
+            "yarn" => Ok(Self::Yarn),
+            "linear" => Ok(Self::Linear),
+            "dynamic" => Ok(Self::Dynamic),
+            _ => Err(candle::Error::Msg(
```
I've removed this `impl` entirely

EricLBuehler commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1114 @@
+    ) -> Result<Tensor> {
+        let (bs, seq_len, _) = xs.dims3()?;
+
+        let mut q = self.q.forward(xs)?;
```
I switched to using local scopes where possible; however, I think some are necessary (rope, kv cache, etc).

EricLBuehler commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1114 @@
+            q_pe.dtype(),
+            q_pe.device(),
+        )?;
+        q = q.slice_assign(
```
Ah, nice catch! I think it would be D::Minus1 though?

EricLBuehler commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1114 @@
+}
+
+pub trait SplitOp {
+    fn split<D: Dim>(&self, splits: &[usize], dim: D) -> Result<Vec<Tensor>>;
```
Actually, my idea was to maybe eventually migrate all the `...Op` traits in this file to `candle_nn` as they are both used in DeepSeek 3 (that PR duplicates these ops) and might be very useful.

What do you think?

LaurentMazare replied:
I would rather keep `candle_nn` on the simpler side, here it seems that the ops can be just specialized to what is exactly required by the model, I would even suggest defining helper functions rather than going through traits, and the deepseek-v3 implementation can use some functions from the deepseek-v2 module.

EricLBuehler replied:
Sounds great, let's keep this as-is then.

LaurentMazare commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1053 @@
+                // (n, topk_group)
+                let group_idx = scores.topk_unsorted(self.cfg.topk_group)?.indices;
+                // (n, n_group)
+                let mut group_mask = group_scores.zeros_like()?;
```
Cannot you just avoid this `mut` by chaining calls or using a local scope, seems fairly easy to do. Please also review the other remaining muts.

EricLBuehler replied:
I've removed this `mut`. I removed all other `mut` occurrences in the model execution path (unless completely impossible), too.

guoqingbao commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1053 @@
+        (q_pe, k_pe) = self.rotary_emb.forward(&q_pe, &k_pe, seqlen_offset)?;
+
+        let q = Tensor::cat(&[q_nope, q_pe], D::Minus1)?;
+        let mut k = Tensor::cat(&[k_nope, k_pe], D::Minus1)?;
```
Hi @EricLBuehler I got the following error for running DeepSeek-V2-Lite-Chat model:
```
called `Result::unwrap()` on an `Err` value: APIError { data: "shape mismatch in cat for dim 1, shape for arg 1: [1, 16, 5, 128] shape for arg 2: [1, 1, 5, 64]" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I found that you have repeated the second dim of Tensor `k_pe` to match Tensor `k_nope` and padded Tensor `v` with zeros in mistral.rs but not over here. Are there something specials for this case? I also tested the mistral.rs implementation, while, the lite model gives random outputs.

EricLBuehler replied:
Hi @guoqingbao!

> I found that you have repeated the second dim of Tensor k_pe to match Tensor k_nope and padded Tensor v with zeros in mistral.rs but not over here. Are there something specials for this case?

That case is only for the v head dim to match q/k as PagedAttention requires that (be sure to unpad too). We don't have PagedAttention in Candle (yet?), so this is not included.

> I also tested the mistral.rs implementation, while, the lite model gives random outputs.

Just checked it there with DS V2 Lite on Metal in mistral.rs, and it works. Was that failing?

guoqingbao commented on candle-transformers/src/models/deepseek2.rs:
```diff
@@ -0,0 +1,1053 @@
+        (q_pe, k_pe) = self.rotary_emb.forward(&q_pe, &k_pe, seqlen_offset)?;
+
+        let q = Tensor::cat(&[q_nope, q_pe], D::Minus1)?;
+        let mut k = Tensor::cat(&[k_nope, k_pe], D::Minus1)?;
```
It can simply fix the problem by adding a dim repeat

```
let k_pe = k_pe.repeat((1, q.dim(1)?, 1, 1))?;
```

EricLBuehler replied:
Fixed that, thanks for finding this!

EricLBuehler commented:
@LaurentMazare I updated the model with some fixes and removed all the `mut` occurrences, except for in the MoE block where it is necessary. If you could review again, that would be great!

I tested the model, and it is working:
```
cargo run --example deepseekv2 --release --features metal -- --prompt "Recursive fibonacci code in Rust:" --which lite --sample-len 150   
```

EricLBuehler commented:
Thank you! I'll update the DeepSeek V3 PR #2745.

I also have some MoE-specific optimizations which I'll be pushing shortly!
